components/remerge/src/storage/db.rs

```diff
@@ -143,16 +122,238 @@ impl RemergeDb {
+            FROM rec_mirror
+            WHERE guid = :guid
+        ";
+        Ok(self
+            .db
+            .execute_named_cached(sql, named_params! { ":guid": guid })?)
+    }
+
+    fn mark_mirror_overridden(&self, guid: &str) -> Result<()> {
+        self.db.execute_named_cached(
+            "UPDATE rec_mirror SET is_overridden = 1 WHERE guid = :guid",
+            named_params! { ":guid": guid },
+        )?;
+        Ok(())
+    }
+
+    /// Combines get_vclock with counter_bump, and produces a new VClock with the bumped counter.
+    fn get_bumped_vclock(&self, id: &str) -> Result<VClock> {
+        let vc = self.get_vclock(id)?;
+        let counter = self.counter_bump()?;
+        Ok(vc.apply(self.client_id.clone(), counter))
+    }
+
+    pub fn update_record(&self, record: &NativeRecord) -> Result<()> {
+        let (guid, record) = self.info.native_to_local(record, ToLocalReason::Update)?;
+        let tx = self.db.unchecked_transaction()?;
+        if self.dupe_exists(&record)? {
+            throw!(InvalidRecord::Duplicate);
+        }
+
+        // Note: These fail with NoSuchRecord if the record doesn't exist.
+        self.ensure_local_overlay_exists(guid.as_str())?;
+        self.mark_mirror_overridden(guid.as_str())?;
+
+        let now_ms = MsTime::now();
+
+        let vclock = self.get_bumped_vclock(&guid)?;
+
+        let sql = "
+            UPDATE rec_local
+            SET local_modified_ms      = :now_millis,
+                record_data            = :record,
+                vector_clock           = :vclock,
+                last_writer_id         = :own_id,
+                remerge_schema_version = :schema_ver,
+                sync_status            = max(sync_status, :changed)
+            WHERE guid = :guid
+        ";
+
+        self.db.execute_named(
+            &sql,
+            named_params! {
+                ":guid": guid,
+                ":changed": SyncStatus::Changed as u8,
+                ":record": record,
+                ":schema_ver": self.info.local.version.to_string(),
+                ":now_millis": now_ms,
+                ":own_id": self.client_id,
+                ":vclock": vclock,
+            },
+        )?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    pub fn client_id(&self) -> Guid {
+        // Guid are essentially free unless the Guid ends up in the "large guid"
+        // path, which should never happen for remerge client ids, so it should
+        // be fine to always clone this.
+        self.client_id.clone()
+    }
+
+    pub fn info(&self) -> &RemergeInfo {
+        &self.info
+    }
+
+    fn dupe_exists(&self, _record: &LocalRecord) -> Result<bool> {
+        // XXX FIXME: this is obviously wrong, but should work for
+        // extension-storage / engines that don't do deduping. (Is it correct
+        // that ext-storage won't want to dedupe on anything?)
```

Comment:
Yes, we don't do deduping in the Kinto implementation of `chrome.storage.sync`, instead choosing IDs carefully so that they match corresponding records on other clients without having to coordinate.

In the implementation of `chrome.storage.sync`, a Kinto bucket corresponds to a user's "slice" of the sync server, each extension is synchronized to a collection, and each key/value pair is stored as a record (with fields `id`, `key`, and `data`, which is whatever the value of the preference is). I'm not 100% sure how those map to Sync 1.5 concepts (it seems like a Sync 1.5 collection corresponds to exactly 1 "engine") or how records will be structured in the Remerge world (see [my comment here](https://github.com/mozilla/application-services/pull/658#pullrequestreview-323929136)). The structure we used in the Kinto implementation was motivated by conflict resolution: tombstones are per-record in Kinto, and it seemed correct to resolve conflicts on a per-preference level rather than per-extension.

Reply:
> The structure we used in the Kinto implementation was motivated by conflict resolution: tombstones are per-record in Kinto, and it seemed correct to resolve conflicts on a per-preference level rather than per-extension.

That makes sense, thanks! `untyped_map` supports full merging, including tombstones, so it sounds like we can preserve the same semantics when we implement it!

> * If the field is an `untyped_map`: performs a similar operation to the 3WM where deltas are created, and each field is merged by `take_newest`.
>   - fIf `prefer_deletions` is true, then if one field is set to the tombstone value, delete it.